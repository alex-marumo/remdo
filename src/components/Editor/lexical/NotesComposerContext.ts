import { NotesState } from "./api";
import { FULL_RECONCILE } from "@lexical/LexicalConstants";
import {
  LexicalComposerContextType,
  useLexicalComposerContext,
} from "@lexical/react/LexicalComposerContext";
import { LexicalEditor } from "lexical";
import { EditorUpdateOptions } from "lexical/LexicalEditor";

export interface NotesLexicalEditor extends LexicalEditor {
  fullUpdate(updateFunction: () => void, options?: EditorUpdateOptions): void;
}

export function useNotesLexicalComposerContext() {
  const [editor, context] = useLexicalComposerContext() as [
    NotesLexicalEditor,
    LexicalComposerContextType
  ];
  editor.fullUpdate = (updateFunction, options) => {
    editor._dirtyType = FULL_RECONCILE;
    editor.update(() => {
      NotesState.getActive()._forceLexicalUpdate();
      updateFunction();
    }, options);
  };
  return [editor, context] as [NotesLexicalEditor, LexicalComposerContextType];
}
