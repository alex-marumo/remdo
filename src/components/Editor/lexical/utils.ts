import { $getNearestNodeFromDOMNode, $isDecoratorNode } from "lexical";

//copied, unexported function from lexical/packages/lexical-rich-text/src/index.ts
export function $isTargetWithinDecorator(target: HTMLElement): boolean {
  const node = $getNearestNodeFromDOMNode(target);
  return $isDecoratorNode(node);
}
