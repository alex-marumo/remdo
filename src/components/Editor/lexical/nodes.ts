import { NotesState, Note, getNotesEditorState } from "./api";
import { patch } from "@/utils";
import { $isListNode, ListItemNode } from "@lexical/list";
import { addClassNamesToElement } from "@lexical/utils";
import { EditorConfig, NodeKey } from "lexical";
import { LexicalNode } from "lexical";


export function applyNodePatches(NodeType: any) {
  /*
    This function customizes updateDOM and createDOM (see below) in an existing
    lexical node class. An alternative would be to use lexical node replacement
    mechanism, but this turns off TextNode merging (see TextNote.isSimpleText()) 
    This could be fixed, but an additional benefit is that this method is 
    simpler, shorter (doesn't require to implement importJSON, clone, etc.)
    and doesn't rename original types
    */
  patch(NodeType, "updateDOM", function (oldMethod, prevNode, dom, config) {
    //lexicalMethod has to be placed first as it may have some side effects
    return (
      //TODO perform an actual check
      oldMethod(prevNode, dom, config) || true
    );
  });

  patch(NodeType, "createDOM", function (oldMethod, config, editor) {
    const notesState = NotesState.getActive();
    //
    // search filter
    //
    //TODO try to use "changeFocus" tag instead of changing state
    if (notesState.filter) {
      if ($isListNode(this)) {
        //TODO this could be specific for ListNode.createDOM
        if (this.getParent().getKey() === "root") {
          return oldMethod(config, editor);
        }
        return document.createElement("div");
      }
      if (!Note.from(this).text.includes(notesState.filter)) {
        return document.createElement("div");
      } else {
        //during search focus and fold are ignored
        return oldMethod(config, editor);
      }
    }

    //
    // focus & fold
    //
    //TODO simplify
    const note = Note.from(this);
    const parents = [...note.parents];
    if (
      !notesState.focus ||
      notesState.focus.parentKey === this.getKey() ||
      notesState.focus.nodeKey === this.getKey() ||
      [note, ...note.parents].some(
        p => p.lexicalKey === notesState.focus.nodeKey
      )
    ) {
      //
      // is fold?
      //
      if (notesState?.focus?.nodeKey !== note.lexicalKey) {
        for (const p of parents) {
          if (p.fold) {
            return document.createElement("div");
          }
          if (p.lexicalKey === notesState.focus?.nodeKey) {
            break;
          }
        }
      }
      const dom: HTMLElement = oldMethod(config, editor);
      if (note.fold) {
        addClassNamesToElement(dom, "note-folded");
      }
      return dom;
    } else {
      return document.createElement("div");
    }
  });
}

//TODO can't use patch because it's static
const oldClone = ListItemNode.clone;
ListItemNode.clone = function(oldNode: ListItemNode) {
  const newNode = oldClone(oldNode);
  newNode.__fold = oldNode.__fold ?? false;
  return newNode;
}

//can't use patch because it's static
const oldImportJSON = ListItemNode.importJSON;
ListItemNode.importJSON = function(serializedNode) {
  const node = oldImportJSON(serializedNode);
  node.__fold = serializedNode.fold;
  return node;
};

patch(ListItemNode, "exportJSON", function(oldExportJSON) {
  return {
    ...oldExportJSON(),
    fold: this.__fold
  }
});
