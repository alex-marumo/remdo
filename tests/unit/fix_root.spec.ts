import { debug, getMinimizedState, loadEditorState } from "./common";
import { $createListItemNode, $createListNode } from "@lexical/list";
import { $createParagraphNode, $getRoot } from "lexical";
import { it } from "vitest";

it("empty", async ({ editor, expect, lexicalUpdate }) => {
  const state = getMinimizedState(editor);
  expect(state).toMatchSnapshot();
});

it("single list - correct", async ({ editor, expect, lexicalUpdate }) => {
  lexicalUpdate(() => {
    const root = $getRoot();
    //root is cleared already in beforeEach, but after that it's the tested root
    //node transform listener is called, so we need to clear it again
    root.clear();
    const listNode = $createListNode("bullet");
    const listItemNode = $createListItemNode();
    listNode.append(listItemNode);
    root.append(listNode);
  });
  const state = getMinimizedState(editor);
  expect(state).toMatchSnapshot();
});

it("multiple lists", async ({ editor, expect, lexicalUpdate }) => {
  lexicalUpdate(() => {
    const root = $getRoot();
    //in total 3 lists as one is added automatically by the tested root node 
    //transform listener
    for (let i = 0; i < 2; i++) {
      const listNode = $createListNode("bullet");
      const listItemNode = $createListItemNode();
      listNode.append(listItemNode);
      root.append(listNode);
    }
  });
  const state = getMinimizedState(editor);
  expect(state).toMatchSnapshot();
});
