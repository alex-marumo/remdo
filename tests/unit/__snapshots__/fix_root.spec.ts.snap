// Vitest Snapshot v1

exports[`empty 1`] = `
"root:
  type: root
  children:
  - type: list
    children:
    - type: listitem
      children: []
"
`;

exports[`multiple lists 1`] = `
"root:
  type: root
  children:
  - type: list
    children:
    - type: listitem
      children: []
    - type: listitem
      children: []
    - type: listitem
      children: []
"
`;

exports[`single list - correct 1`] = `
"root:
  type: root
  children:
  - type: list
    children:
    - type: listitem
      children: []
"
`;
